empty elementOfJsonPath

vegterb opened the pull request:
return null with empty elementOfJsonPath

fhoeben commented:
Looks good. Would you mind adding a test?

vegterb commented:
JsonFixtureTest Class & twee testen aangemaakt

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/JsonFixture.java:
```diff
@@ -73,7 +73,7 @@ public String createFileFromBase64ContentOf(String baseName, String jsonPath) {
     public Object elementOfJsonPath(int index, String path) {
         List<Object> all = listJsonPathMatches(path);
-        return all.get(index);
+        return all.size() > 0 ? all.get(index) : null;
```
```suggestion
        return all.size() > index ? all.get(index) : null;
```

vegterb replied:
Ah, yes... Interesting. This would make sure that when you chose an element which does not exist.

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    public void testElementOfJsonPathNoResult() {
+        JsonFixture cont = new JsonFixture();
```
Why not just use `fixture`?

vegterb replied:
Agreed

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    public void testElementOfJsonPathResults(){
+        JsonFixture cont = new JsonFixture();
```
Why not just use `fixture`?

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+    public void testElementOfJsonPathNoResult() {
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathNoResults);
```
```suggestion
        Object result = cont.elementOfJsonPath(0, jsonPathNoResults);
```

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+    public void testElementOfJsonPathResults(){
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathWithResults);
```
```suggestion
        Object result = cont.elementOfJsonPath(0, "$.store.book[?(@.category==\"fiction\")].author");
```

vegterb replied:
Interesting, why would you keep these values in code and not variables? Does this improve speed? It looked cleaner with the variable, but either way is fine with me.

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathWithResults);
+        assertEquals(expectedResult,result);
```
```suggestion
        assertEquals(expectedResult, "Evelyn Waugh");
```

vegterb replied:
Blindly added, but I expect you wanted to replace the expectedResult variable here.

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+    private JsonFixture fixture = new JsonFixture();
+    private final String jsonPathNoResults = "$.store.book[?(@.category==\"test\")].author";
+    private final String jsonPathWithResults = "$.store.book[?(@.category==\"fiction\")].author";
+    private final String expectedResult = "Evelyn Waugh";
```
remove

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+public class JsonFixtureTest {
+    private JsonFixture fixture = new JsonFixture();
+    private final String jsonPathNoResults = "$.store.book[?(@.category==\"test\")].author";
+    private final String jsonPathWithResults = "$.store.book[?(@.category==\"fiction\")].author";
```
remove

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:
```diff
@@ -0,0 +1,74 @@
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathNoResults);
+        assertNull(result);
```
Add test that getting element 1 also returns `null`

vegterb replied:
Add test with no result, where the jsonPath results in 3 items, but item 3 returns null.

fhoeben reviewed (approved):
Looks good. A few minor points to finish it I believe

vegterb commented:
@fhoeben changes made, thanks for the feedback.

fhoeben commented:
Just remembered `JsonHttpTest`: has its own handling (if only Java had mixins) of getting elements. Please make the same change there. 
Sorry for not mentioning it earlier...

vegterb commented:
@fhoeben as per your comment I updated `JsonHttpTest` but got a bit stuck on the unit test, my girlfriend suggested Mockito, which was already part of your project so I also updated the JsonFixtureTest. Let me know what you think.

fhoeben commented:
Although I really like mockito, I try to keep away from using spies as much as possible as they tend to couple your test too tightly to the implementation. Instead I suggest you take a look at wiki/FitNesseRoot/HsacAcceptanceTests/SlimTests/HttpTest/JsonHttpTest.wiki line 30, I believe adding one line there to check element 2 would be the easiest way to test the changed behaviour (line 30 tests the behaviour when the element _is_ present)

vegterb commented:
I could of course not use the spy implementation and just use a mock for the `HttpResponse` class and keep using all the methodes within the 'JsonHttpTest' class. It's you're call... mock or revert and use:
`|check        |element              |2                   |of json path|airplanes[?(@.airline=='KLM')].type|null|`
is also no problem.

I will revert the changes in the 'JsonFixtureTest'.

fhoeben commented:
I would prefer `|check |element |2 |of json path|airplanes[?(@.airline=='KLM')].type|null|` 

The spy indeed isolates the specific method, but that method is the implementation detail. The FitNesse test checks the behaviour of the class given a certain response from the remote server, it's completely decoupled from how the fixture arrives at that response. So JsonHttpTest could be completely refactored (for instance by delegating to JsonFixture, or even rewriting it in Kotlin) and the test needs no change: it still verifies whether the correct response is given.

vegterb commented:
@fhoeben all changes made, should be good now.

fhoeben commented on src/test/java/nl/hsac/fitnesse/fixture/slim/JsonHttpTestTest.java:
```diff
@@ -6,8 +6,7 @@
 
```
Please revert all changes to this file (`*` imports are a bad idea, and I like newlines at the end of my files)

fhoeben reviewed (changes requested):
Looks good. One minor thing left
